<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Fast and efficient summarization of generic bedGraph files from Bisufite sequencing • methrix</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Fast and efficient summarization of generic bedGraph files from Bisufite sequencing">
<meta property="og:description" content="Bedgraph files generated by Bisulfite pipelines often come in various flavors. Critical downstream step requires summarization of these files into methylation/coverage matrices. This step of data aggregation is done by Methrix, including many other useful downstream functions.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">methrix</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.3.20</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/methrix.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/02-Import.html">01: Reading bedgraph files</a>
    </li>
    <li>
      <a href="articles/03-QC_reports.html">02: QC and html reports</a>
    </li>
    <li>
      <a href="articles/04-DMRs.html">03: Indentifying DMRs</a>
    </li>
    <li>
      <a href="articles/05-Large_cohorts.html">04: Working with large cohorts</a>
    </li>
    <li>
      <a href="articles/06-Export.html">05: Exporting to standard file formats</a>
    </li>
    <li>
      <a href="articles/07-Reader_FAQ.html">FAQ: Detailed documentation of the read_bedgraph function.</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<p><img src="articles/logo_large_hexagon.gif" height="135" width="135" style="align:center"></p>
<hr>
<div id="fast-and-efficient-summarization-of-generic-bedgraph-files-from-bisufite-sequencing" class="section level2">
<h2 class="hasAnchor">
<a href="#fast-and-efficient-summarization-of-generic-bedgraph-files-from-bisufite-sequencing" class="anchor"></a>Fast and efficient summarization of generic bedGraph files from Bisufite sequencing</h2>
<!-- badges: start -->

<!-- badges: end -->
<div id="introduction" class="section level3">
<h3 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h3>
<p>Bedgraph files generated by BS pipelines often come in various flavors. Critical downstream step requires aggregation of these files into methylation/coverage matrices. This step of data aggregation is done by Methrix, including many other useful downstream functions.</p>
</div>
<div id="package-overview-and-usage-functions" class="section level3">
<h3 class="hasAnchor">
<a href="#package-overview-and-usage-functions" class="anchor"></a>Package overview and usage functions</h3>
<p>For more detailed documentation, see the <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/methrix/inst/doc/methrix.html">vignette</a>.</p>
<p>A examplary complete data analysis with steps from reading in to annotation and differential methylation calling can be find in our <a href="https://compepigen.github.io/WGBS_best_practice/best_practices.html">best practice pipeline</a>.</p>
<p align="center">
<img src="https://github.com/CompEpigen/methrix/blob/master/vignettes/overview.png"></p>
</div>
<div id="summary" class="section level3">
<h3 class="hasAnchor">
<a href="#summary" class="anchor"></a>Summary:</h3>
<ul>
<li>Faster summarization of generic bedGraph files with <code>data.table</code> back-end</li>
<li>Fills missing CpGs from reference genome</li>
<li>Vectorized code (faster, memory expensive) and non-vectorized code (slower, minimal memory)</li>
<li>Built upon <code>SummarizedExperiment</code> with custom methods for CpG extraction, sub-setting, and filtering</li>
<li>Easy conversion to bsseq object for downstream analysis</li>
<li>Extensive one click interactive html report generation</li>
<li>Supports serialized arrays with <code>HDF5Array</code> and <code>saveHDF5SummarizedExperiment</code>
</li>
</ul>
</div>
<div id="updates" class="section level3">
<h3 class="hasAnchor">
<a href="#updates" class="anchor"></a>Updates:</h3>
<p>see <a href="https://github.com/CompEpigen/methrix/blob/master/NEWS.md">here</a></p>
</div>
<div id="installation" class="section level3">
<h3 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h3>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"CompEpigen/methrix"</span><span class="op">)</span></pre></div>
</div>
<div id="usage" class="section level3">
<h3 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage:</h3>
<div id="step-1-extract-all-cpg-loci-from-the-reference-genome" class="section level4">
<h4 class="hasAnchor">
<a href="#step-1-extract-all-cpg-loci-from-the-reference-genome" class="anchor"></a>Step-1: Extract all CpG loci from the reference genome</h4>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="op">&gt;</span><span class="st"> </span>hg19_cpgs =<span class="st"> </span>methrix<span class="op">::</span><span class="kw">extract_CPGs</span>(<span class="dt">ref_genome =</span> <span class="st">"BSgenome.Hsapiens.UCSC.hg19"</span>)</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="op">-</span>Extracting CpGs</span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="op">-</span>Done. Extracted <span class="dv">29</span>,<span class="dv">891</span>,<span class="dv">155</span> CpGs from <span class="dv">298</span> contigs.</span>
<span id="cb2-4"><a href="#cb2-4"></a>There were <span class="dv">50</span> or more <span class="kw">warnings</span> (use <span class="kw">warnings</span>() to see the first <span class="dv">50</span>)</span></code></pre></div>
</div>
<div id="step-2-read-in-bedgraphs-and-generate-a-methrix-object" class="section level4">
<h4 class="hasAnchor">
<a href="#step-2-read-in-bedgraphs-and-generate-a-methrix-object" class="anchor"></a>Step-2: Read in bedgraphs and generate a methrix object</h4>
<p>The example data of the methrix package is used.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="co">#Example bedgraph files</span></span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="op">&gt;</span><span class="st"> </span>bdg_files =<span class="st"> </span><span class="kw">list.files</span>(<span class="dt">path =</span> <span class="kw">system.file</span>(<span class="st">'extdata'</span>, <span class="dt">package =</span> <span class="st">'methrix'</span>), <span class="dt">pattern =</span> <span class="st">"*bdg</span><span class="ch">\\</span><span class="st">.gz$"</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>)</span>
<span id="cb3-3"><a href="#cb3-3"></a></span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="op">&gt;</span><span class="st"> </span>meth =<span class="st"> </span>methrix<span class="op">::</span><span class="kw">read_bedgraphs</span>(<span class="dt">files =</span> bdg_files, <span class="dt">ref_cpgs =</span> hg19_cpgs, <span class="dt">chr_idx =</span> <span class="dv">1</span>, <span class="dt">start_idx =</span> <span class="dv">2</span>, <span class="dt">M_idx =</span> <span class="dv">3</span>, <span class="dt">U_idx =</span> <span class="dv">4</span>,</span>
<span id="cb3-5"><a href="#cb3-5"></a>  <span class="dt">stranded =</span> <span class="ot">TRUE</span>, <span class="dt">collapse_strands =</span> <span class="ot">TRUE</span>)</span>
<span id="cb3-6"><a href="#cb3-6"></a><span class="op">----------------------------</span></span>
<span id="cb3-7"><a href="#cb3-7"></a><span class="op">-</span>Preset<span class="op">:</span><span class="st">        </span>Custom</span>
<span id="cb3-8"><a href="#cb3-8"></a><span class="op">--</span>Missing beta and coverage info. Estimating them from M and U values</span>
<span id="cb3-9"><a href="#cb3-9"></a><span class="op">-</span>CpGs raw<span class="op">:</span><span class="st">      </span><span class="dv">29</span>,<span class="dv">891</span>,<span class="dv">155</span> (total reference CpGs)</span>
<span id="cb3-10"><a href="#cb3-10"></a><span class="op">-</span>CpGs retained<span class="op">:</span><span class="st"> </span><span class="dv">28</span>,<span class="dv">217</span>,<span class="dv">448</span>(reference CpGs from contigs of interest)</span>
<span id="cb3-11"><a href="#cb3-11"></a><span class="op">-</span>CpGs stranded<span class="op">:</span><span class="st"> </span><span class="dv">56</span>,<span class="dv">434</span>,<span class="dv">896</span>(reference CpGs from both strands)</span>
<span id="cb3-12"><a href="#cb3-12"></a><span class="op">----------------------------</span></span>
<span id="cb3-13"><a href="#cb3-13"></a><span class="op">-</span>Processing<span class="op">:</span><span class="st">    </span>C1.bedGraph.gz</span>
<span id="cb3-14"><a href="#cb3-14"></a><span class="op">--</span>CpGs missing<span class="op">:</span><span class="st"> </span><span class="dv">56</span>,<span class="dv">434</span>,<span class="dv">219</span> (from known reference CpGs)</span>
<span id="cb3-15"><a href="#cb3-15"></a><span class="op">-</span>Processing<span class="op">:</span><span class="st">    </span>C2.bedGraph.gz</span>
<span id="cb3-16"><a href="#cb3-16"></a><span class="op">--</span>CpGs missing<span class="op">:</span><span class="st"> </span><span class="dv">56</span>,<span class="dv">434</span>,<span class="dv">207</span> (from known reference CpGs)</span>
<span id="cb3-17"><a href="#cb3-17"></a><span class="op">-</span>Processing<span class="op">:</span><span class="st">    </span>N1.bedGraph.gz</span>
<span id="cb3-18"><a href="#cb3-18"></a><span class="op">--</span>CpGs missing<span class="op">:</span><span class="st"> </span><span class="dv">56</span>,<span class="dv">434</span>,<span class="dv">194</span> (from known reference CpGs)</span>
<span id="cb3-19"><a href="#cb3-19"></a><span class="op">-</span>Processing<span class="op">:</span><span class="st">    </span>N2.bedGraph.gz</span>
<span id="cb3-20"><a href="#cb3-20"></a><span class="op">--</span>CpGs missing<span class="op">:</span><span class="st"> </span><span class="dv">56</span>,<span class="dv">434</span>,<span class="dv">195</span> (from known reference CpGs)</span>
<span id="cb3-21"><a href="#cb3-21"></a><span class="op">-</span>Finished <span class="cf">in</span><span class="op">:</span><span class="st">  </span><span class="dv">00</span><span class="op">:</span><span class="dv">02</span><span class="op">:</span><span class="dv">00</span> <span class="kw">elapsed</span> (<span class="dv">00</span><span class="op">:</span><span class="dv">02</span><span class="op">:</span><span class="dv">23</span> cpu)</span>
<span id="cb3-22"><a href="#cb3-22"></a></span>
<span id="cb3-23"><a href="#cb3-23"></a><span class="op">&gt;</span><span class="st"> </span>meth</span>
<span id="cb3-24"><a href="#cb3-24"></a>An object of class methrix</span>
<span id="cb3-25"><a href="#cb3-25"></a>   n_CpGs<span class="op">:</span><span class="st"> </span><span class="dv">28</span>,<span class="dv">217</span>,<span class="dv">448</span></span>
<span id="cb3-26"><a href="#cb3-26"></a>n_samples<span class="op">:</span><span class="st"> </span><span class="dv">4</span></span>
<span id="cb3-27"><a href="#cb3-27"></a>    is_h5<span class="op">:</span><span class="st"> </span><span class="ot">FALSE</span></span>
<span id="cb3-28"><a href="#cb3-28"></a>Reference<span class="op">:</span><span class="st"> </span>hg19</span></code></pre></div>
</div>
</div>
<div id="methrix-operations" class="section level3">
<h3 class="hasAnchor">
<a href="#methrix-operations" class="anchor"></a>Methrix operations</h3>
<p>What can be done on <code>methrix</code> object? Following are the key functions</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="co">#reading and writing:</span>
<span class="fu"><a href="reference/read_bedgraphs.html">read_bedgraphs</a></span><span class="op">(</span><span class="op">)</span>    <span class="co">#Reads in bedgraph files into methrix</span>
<span class="fu"><a href="reference/write_bedgraphs.html">write_bedgraphs</a></span><span class="op">(</span><span class="op">)</span>   <span class="co">#Writes bedGraphs from methrix object</span>
<span class="fu"><a href="reference/write_bigwigs.html">write_bigwigs</a></span><span class="op">(</span><span class="op">)</span>     <span class="co">#Writes bigWigs from methrix object</span>
<span class="co">#operations</span>
<span class="fu"><a href="reference/order_by_sd.html">order_by_sd</a></span><span class="op">(</span><span class="op">)</span>       <span class="co">#Orders methrix object by SD</span>
<span class="fu"><a href="reference/region_filter.html">region_filter</a></span><span class="op">(</span><span class="op">)</span>     <span class="co">#Filters matrices by region</span>
<span class="fu"><a href="reference/mask_methrix.html">mask_methrix</a></span><span class="op">(</span><span class="op">)</span>      <span class="co">#Masks lowly covered CpGs</span>
<span class="fu"><a href="reference/coverage_filter.html">coverage_filter</a></span><span class="op">(</span><span class="op">)</span>   <span class="co">#Filters methrix object based on coverage</span>
<span class="fu"><a href="reference/subset_methrix.html">subset_methrix</a></span><span class="op">(</span><span class="op">)</span>      <span class="co">#Subsets methrix object based on given conditions.</span>
<span class="fu"><a href="reference/remove_uncovered.html">remove_uncovered</a></span><span class="op">(</span><span class="op">)</span>  <span class="co">#Removes loci that are uncovered across all samples</span>
<span class="fu"><a href="reference/remove_snps.html">remove_snps</a></span><span class="op">(</span><span class="op">)</span>       <span class="co">#Removes loci overlapping with possible SNPs</span>
<span class="co">#Visualization and QC</span>
<span class="fu"><a href="reference/methrix_report.html">methrix_report</a></span><span class="op">(</span><span class="op">)</span>    <span class="co">#Creates a detailed interative html summary report from methrix object</span>
<span class="fu"><a href="reference/methrix_pca.html">methrix_pca</a></span><span class="op">(</span><span class="op">)</span>         <span class="co">#Principal Component Analysis</span>
<span class="fu"><a href="reference/plot_pca.html">plot_pca</a></span><span class="op">(</span><span class="op">)</span>          <span class="co">#Plots the result of PCA</span>
<span class="fu"><a href="reference/plot_coverage.html">plot_coverage</a></span><span class="op">(</span><span class="op">)</span>     <span class="co">#Plots coverage statistics</span>
<span class="fu"><a href="reference/plot_density.html">plot_density</a></span><span class="op">(</span><span class="op">)</span>      <span class="co">#Plots the density distribution of the beta values </span>
<span class="fu"><a href="reference/plot_violin.html">plot_violin</a></span><span class="op">(</span><span class="op">)</span>       <span class="co">#Plots the distribution of the beta values on a violin plot</span>
<span class="fu"><a href="reference/plot_stats.html">plot_stats</a></span><span class="op">(</span><span class="op">)</span>        <span class="co">#Plot descriptive statistics</span>
<span class="fu"><a href="reference/get_stats.html">get_stats</a></span><span class="op">(</span><span class="op">)</span>         <span class="co">#Estimate descriptive statistics of the object</span>
<span class="co">#Other</span>
<span class="fu"><a href="reference/methrix2bsseq.html">methrix2bsseq</a></span><span class="op">(</span><span class="op">)</span>     <span class="co">#Convert methrix to bsseq object</span></pre></div>
</div>
<div id="note" class="section level3">
<h3 class="hasAnchor">
<a href="#note" class="anchor"></a>Note</h3>
<p><strong><em>This repository is under active development</em></strong></p>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from BIOC at <br><a href="https://www.bioconductor.org/packages/methrix">https://​www.bioconductor.org/​packages/​methrix</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Anand Mayakonda <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-1162-687X" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Reka Toth <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-6096-1052" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.tidyverse.org/lifecycle/#maturing"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle: maturing"></a></li>
<li><a href="https://bioconductor.org/checkResults/release/bioc-LATEST/methrix"><img src="http://www.bioconductor.org/shields/build/devel/bioc/methrix.svg" alt="BioC status"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Anand Mayakonda, Reka Toth.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
