<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>03: Indentifying DMRs • methrix</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="03: Indentifying DMRs">
<meta property="og:description" content="methrix">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">methrix</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.3.20</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/methrix.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/02-Import.html">01: Reading bedgraph files</a>
    </li>
    <li>
      <a href="../articles/03-QC_reports.html">02: QC and html reports</a>
    </li>
    <li>
      <a href="../articles/04-DMRs.html">03: Indentifying DMRs</a>
    </li>
    <li>
      <a href="../articles/05-Large_cohorts.html">04: Working with large cohorts</a>
    </li>
    <li>
      <a href="../articles/06-Export.html">05: Exporting to standard file formats</a>
    </li>
    <li>
      <a href="../articles/07-Reader_FAQ.html">FAQ: Detailed documentation of the read_bedgraph function.</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="04-DMRs_files/accessible-code-block-0.0.1/empty-anchor.js"></script><script src="04-DMRs_files/htmlwidgets-1.5.1/htmlwidgets.js"></script><script src="04-DMRs_files/plotly-binding-4.9.2.1/plotly.js"></script><script src="04-DMRs_files/typedarray-0.1/typedarray.min.js"></script><script src="04-DMRs_files/jquery-1.11.3/jquery.min.js"></script><link href="04-DMRs_files/crosstalk-1.1.0.1/css/crosstalk.css" rel="stylesheet">
<script src="04-DMRs_files/crosstalk-1.1.0.1/js/crosstalk.min.js"></script><link href="04-DMRs_files/plotly-htmlwidgets-css-1.52.2/plotly-htmlwidgets.css" rel="stylesheet">
<script src="04-DMRs_files/plotly-main-1.52.2/plotly-latest.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>03: Indentifying DMRs</h1>
            
      
      
      <div class="hidden name"><code>04-DMRs.Rmd</code></div>

    </div>

    
    
<div id="differential-methylation-calling" class="section level1">
<h1 class="hasAnchor">
<a href="#differential-methylation-calling" class="anchor"></a>Differential methylation calling</h1>
<p>Identifying Deferentially methylated regions (DMRs) can be done using within R or using external command line programs. Here we describe how methrix object can be used with both the approaches. For DMR calling in R, popular programs include <a href="https://bioconductor.org/packages/release/bioc/vignettes/DSS/inst/doc/DSS.html">DSS</a>, <a href="https://bioconductor.org/packages/release/bioc/vignettes/dmrseq/inst/doc/dmrseq.html">DMRseq</a>. Outside R, <a href="https://www.bioinf.uni-leipzig.de/Software/metilene/">metilene</a> offers a fast and simple alternative. Note that DNRseq and DSS can be resource hungry with run rimes often involving hours. <code>Metilene</code> however is ultrafast and lightweight with small memory footprint.</p>
<div id="dss" class="section level2">
<h2 class="hasAnchor">
<a href="#dss" class="anchor"></a>DSS</h2>
<p>A very important step of a WGBS data analysis is the differential methylation calling. During this step, we would like to identify changes between groups or conditions first on the site level. Then, we will need to identify larger regions (DMR-s, differentially methylated regions) affected by methylation changes that are more likely to represent functional alterations. <code>Methrix</code> doesn’t have a DMR caller, therefore we will use <code>DSS</code>. At this stage, <code>DSS</code> is not yet able to work directly with <code>methrix</code>, therefore we transform our methrix object to <code>bsseq</code> for the sake of DMR calling.</p>
<div class="sourceCode" id="cb1"><pre class="downlit">

<span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"DSS"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"DSS"</span><span class="op">)</span>
<span class="op">}</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">DSS</span><span class="op">)</span></pre></div>
<p>For more detailed description, options and recommendations on smoothing, please refer to the ´DSS´ <a href="https://bioconductor.org/packages/release/bioc/vignettes/DSS/inst/doc/DSS.html">vignette</a></p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="va">bs_obj</span> <span class="op">&lt;-</span> <span class="fu">methrix</span><span class="fu">::</span><span class="fu"><a href="../reference/methrix2bsseq.html">methrix2bsseq</a></span><span class="op">(</span><span class="va">meth</span><span class="op">)</span>
<span class="va">bs_obj</span>
<span class="co">#&gt; An object of type 'BSseq' with</span>
<span class="co">#&gt;   28217448 methylation loci</span>
<span class="co">#&gt;   6 samples</span>
<span class="co">#&gt; has not been smoothed</span>
<span class="co">#&gt; All assays are in-memory</span></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="va">dmlTest</span> <span class="op">=</span> <span class="fu">DSS</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DSS/man/DMLtest.html">DMLtest</a></span><span class="op">(</span><span class="va">bs_obj</span>, group1<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">sample_anno</span><span class="op">)</span><span class="op">[</span><span class="va">sample_anno</span><span class="op">$</span><span class="va">Day</span><span class="op">==</span><span class="st">"D6"</span><span class="op">]</span>, group2<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">sample_anno</span><span class="op">)</span><span class="op">[</span><span class="va">sample_anno</span><span class="op">$</span><span class="va">Day</span><span class="op">==</span><span class="st">"D14"</span><span class="op">]</span>, 
                       smoothing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></pre></div>
<div id="call-dmrs-and-dmls" class="section level3">
<h3 class="hasAnchor">
<a href="#call-dmrs-and-dmls" class="anchor"></a>Call DMRs and DMLs</h3>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="va">dmls</span> <span class="op">=</span> <span class="fu">DSS</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DSS/man/callDML.html">callDML</a></span><span class="op">(</span><span class="va">dmlTest</span>, p.threshold <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span>
<span class="va">dmrs</span> <span class="op">=</span> <span class="fu">DSS</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DSS/man/callDMR.html">callDMR</a></span><span class="op">(</span><span class="va">dmlTest</span>, p.threshold <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></pre></div>
<p><strong>DMRs:</strong></p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">dmrs</span><span class="op">)</span>
<span class="co">#&gt;          chr     start       end length nCG meanMethy1 meanMethy2 diff.Methy</span>
<span class="co">#&gt; 292302  chr7   1686521   1687244    724  36  0.9253184  0.4316565  0.4936619</span>
<span class="co">#&gt; 120729 chr16  73582841  73585140   2300  28  0.8963596  0.3711405  0.5252191</span>
<span class="co">#&gt; 91590  chr13 112225194 112225927    734  22  0.9684702  0.3701374  0.5983328</span>
<span class="co">#&gt; 122302 chr16  85482141  85483212   1072  41  0.7926689  0.4123773  0.3802916</span>
<span class="co">#&gt; 897     chr1  10732115  10733515   1401  37  0.9412203  0.5546301  0.3865901</span>
<span class="co">#&gt; 195240  chr3  18411127  18413011   1885  25  0.9463775  0.4042538  0.5421237</span>
<span class="co">#&gt;        areaStat</span>
<span class="co">#&gt; 292302 258.9261</span>
<span class="co">#&gt; 120729 224.1427</span>
<span class="co">#&gt; 91590  220.4902</span>
<span class="co">#&gt; 122302 201.0486</span>
<span class="co">#&gt; 897    196.7819</span>
<span class="co">#&gt; 195240 191.6146</span></pre></div>
<p><strong>DMLs:</strong></p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">dmls</span><span class="op">)</span>
<span class="co">#&gt;         chr      pos       mu1        mu2      diff    diff.se     stat</span>
<span class="co">#&gt; 115002 chr1  7164841 0.9478878 0.20845803 0.7394297 0.05289578 13.97899</span>
<span class="co">#&gt; 238604 chr1 17112454 0.9218719 0.30521631 0.6166556 0.06054654 10.18482</span>
<span class="co">#&gt; 278735 chr1 20577038 0.8532530 0.06043134 0.7928217 0.04640743 17.08394</span>
<span class="co">#&gt; 342292 chr1 25803237 0.8620008 0.26811728 0.5938835 0.05771677 10.28962</span>
<span class="co">#&gt; 484470 chr1 38802438 0.9382340 0.31429739 0.6239366 0.05476446 11.39309</span>
<span class="co">#&gt; 580547 chr1 48508231 0.8705919 0.26986039 0.6007315 0.05822061 10.31819</span>
<span class="co">#&gt;              phi1       phi2         pval          fdr postprob.overThreshold</span>
<span class="co">#&gt; 115002 0.04792590 0.05572107 2.094283e-44 1.277624e-38                      1</span>
<span class="co">#&gt; 238604 0.04677971 0.05479802 2.317896e-24 1.353868e-19                      1</span>
<span class="co">#&gt; 278735 0.04869763 0.07494907 1.954656e-65 5.365998e-59                      1</span>
<span class="co">#&gt; 342292 0.04616408 0.06169127 7.848714e-25 4.967516e-20                      1</span>
<span class="co">#&gt; 484470 0.05334149 0.06542750 4.526264e-30 6.061307e-25                      1</span>
<span class="co">#&gt; 580547 0.06240985 0.05203124 5.830977e-25 3.825611e-20                      1</span></pre></div>
</div>
</div>
<div id="dmr-calling-with-dmrseq" class="section level2">
<h2 class="hasAnchor">
<a href="#dmr-calling-with-dmrseq" class="anchor"></a>DMR calling with <code>dmrseq</code>
</h2>
<p>A popular <code>R</code> package for DMR calling is <code>dmrseq</code>. For more detailed description of the possibilites, see the <a href="http://bioconductor.org/packages/release/bioc/vignettes/dmrseq/inst/doc/dmrseq.html">vignette</a>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit">

<span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"dmrseq"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"dmrseq"</span><span class="op">)</span>
<span class="op">}</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">dmrseq</span><span class="op">)</span>

<span class="fu">register</span><span class="op">(</span><span class="fu">MulticoreParam</span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>

<span class="va">testCovariate</span> <span class="op">&lt;-</span> <span class="st">"Day"</span>
<span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu">dmrseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dmrseq/man/dmrseq.html">dmrseq</a></span><span class="op">(</span>bs<span class="op">=</span><span class="va">bs_obj</span><span class="op">[</span><span class="fl">240001</span><span class="op">:</span><span class="fl">260000</span>,<span class="op">]</span>,
                  cutoff <span class="op">=</span> <span class="fl">0.05</span>,
                  testCovariate<span class="op">=</span><span class="va">testCovariate</span>, verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></pre></div>
</div>
<div id="dmr-calling-with-metilene" class="section level2">
<h2 class="hasAnchor">
<a href="#dmr-calling-with-metilene" class="anchor"></a>DMR calling with <code>Metilene</code>
</h2>
<p>If you prefer using <a href="https://www.bioinf.uni-leipzig.de/Software/metilene/">Metilene</a> which is significantly faster than R based tools, you can export the <code>methrix</code> object to <code>Metilene</code> input file formats.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span>path <span class="op">=</span> <span class="st">"./metilene/"</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu">methrix</span><span class="fu">::</span><span class="fu"><a href="../reference/write_bedgraphs.html">write_bedgraphs</a></span><span class="op">(</span>m <span class="op">=</span> <span class="va">meth</span>, output_dir <span class="op">=</span> <span class="st">"./metilene/"</span>, rm_NA <span class="op">=</span> <span class="cn">FALSE</span>, metilene <span class="op">=</span> <span class="cn">TRUE</span>,multiBed <span class="op">=</span> <span class="st">"metline_ip"</span>, phenoCol <span class="op">=</span> <span class="st">"Day"</span>, compress <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="co">#----------------------</span>
<span class="co">#*Writing ./metilene//metline_ip.bedGraph.gz </span>
<span class="co">#----------------------  </span></pre></div>
<p>Once the file is generated, you can run <code>metilene</code> which should take less than couple of minutes to finish.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1"></a>$ <span class="ex">./metilene_v0.2-8/metilene</span> -t 4 -a D14_WT -b D6_WT metilene/metline_ip.bedGraph <span class="op">&gt;</span> metilene/D14_vs_D6.tsv</span>
<span id="cb10-2"><a href="#cb10-2"></a></span>
<span id="cb10-3"><a href="#cb10-3"></a><span class="co">#Sort output by q values</span></span>
<span id="cb10-4"><a href="#cb10-4"></a>$ <span class="fu">cat</span> metilene/D14_vs_D6.tsv <span class="kw">|</span> <span class="fu">sort</span> -k 4 <span class="kw">|</span> <span class="fu">head</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="co"># chr6  163235589   163236271   0.00010009  -0.464417   12  2.717e-12   1.7226e-11  0.499   0.96342</span>
<span class="co"># chr16 72935841    72936574    0.00010268  -0.593495   13  9.7322e-13  1.7671e-11  0.31243 0.90592</span>
<span class="co"># chr3  126891865   126892175   0.00010581  -0.548356   13  1.7963e-13  1.8211e-11  0.27462 0.82297</span>
<span class="co"># chr2  13448747    13448946    0.00011202  0.142889    24  1.3844e-13  1.9279e-11  0.99314 0.85025</span>
<span class="co"># chr1  18381792    18381953    0.00011518  -0.464353   10  5.7773e-11  1.9823e-11  0.45378 0.91813</span>
<span class="co"># chr8  38587666    38587922    0.00011689  -0.475570   10  3.8831e-11  2.0118e-11  0.4333  0.90887</span>
<span class="co"># chr1  44030377    44030580    0.00011877  -0.705450   10  3.1796e-11  2.0442e-11  0.11405 0.8195</span>
<span class="co"># chr17 77158527    77158873    0.00012301  -0.488690   10  4.7383e-11  2.117e-11   0.45551 0.9442</span>
<span class="co"># chr17 75588404    75588890    0.00013396  -0.523367   10  3.1796e-11  2.3054e-11  0.412   0.93537</span>
<span class="co"># chr2  222317260   222317626   0.00013438  -0.598000   10  3.5142e-11  2.3128e-11  0.3712  0.9692</span></pre></div>
</div>
</div>
<div id="visualizing-dmrs-and-candidate-regions" class="section level1">
<h1 class="hasAnchor">
<a href="#visualizing-dmrs-and-candidate-regions" class="anchor"></a>Visualizing DMRs and candidate regions</h1>
<div id="visualizing-a-region" class="section level2">
<h2 class="hasAnchor">
<a href="#visualizing-a-region" class="anchor"></a>Visualizing a region</h2>
<p>We can visualize certain regions and/or DMRs using the <a href="https://bioconductor.org/packages/release/bioc/html/Gviz.html">Gviz</a> package. The plotting function is included in the <a href="https://github.com/CompEpigen/WGBS_best_practice/blob/master/best_practices_WGBS.Rmd">best_practices_WGBS.Rmd</a> file.</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="va">dmrs</span> <span class="op">&lt;-</span> <span class="fu">makeGRangesFromDataFrame</span><span class="op">(</span><span class="va">dmrs</span>, keep.extra.columns <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="va">genome</span> <span class="op">&lt;-</span> <span class="st">"hg19"</span>
<span class="va">mart</span> <span class="op">&lt;-</span> <span class="fu">useMart</span><span class="op">(</span>biomart<span class="op">=</span><span class="st">"ENSEMBL_MART_ENSEMBL"</span>, host<span class="op">=</span><span class="st">"feb2014.archive.ensembl.org"</span>, dataset<span class="op">=</span><span class="st">"hsapiens_gene_ensembl"</span><span class="op">)</span>

<span class="va">i</span> <span class="op">&lt;-</span> <span class="fl">5</span>

<span class="co">#candidate_region &lt;- GRanges(paste0(seqlevels(dmrs)[i], ":", start(dmrs)[i], "-", end(dmrs)[i]))</span>
<span class="va">candidate_region</span> <span class="op">&lt;-</span> <span class="va">dmrs</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>
<span class="fu">region_plot</span><span class="op">(</span>m <span class="op">=</span> <span class="va">meth</span>, region<span class="op">=</span><span class="va">candidate_region</span>, mart<span class="op">=</span><span class="va">mart</span>, genome<span class="op">=</span><span class="va">genome</span>, groups<span class="op">=</span><span class="va">meth</span><span class="op">$</span><span class="va">Day</span>, dmrs<span class="op">=</span><span class="va">dmrs</span><span class="op">)</span>
<span class="co">#the region_plot function is available in the best_practices.Rmd file. </span></pre></div>
</div>
<div id="heatmap-of-dmrs" class="section level2">
<h2 class="hasAnchor">
<a href="#heatmap-of-dmrs" class="anchor"></a>Heatmap of DMRs</h2>
<div class="sourceCode" id="cb14"><pre class="downlit">

<span class="va">heatmap_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu">methrix</span><span class="fu">::</span><span class="fu"><a href="../reference/get_region_summary.html">get_region_summary</a></span><span class="op">(</span><span class="va">meth</span>, <span class="fu">makeGRangesFromDataFrame</span><span class="op">(</span><span class="va">dmrs</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">heatmap_data</span> <span class="op">&lt;-</span> <span class="va">heatmap_data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/stats/complete.cases.html">complete.cases</a></span><span class="op">(</span><span class="va">heatmap_data</span><span class="op">)</span>,<span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">heatmap_data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">heatmap_data</span><span class="op">$</span><span class="va">chr</span>,<span class="st">":"</span>, <span class="va">heatmap_data</span><span class="op">$</span><span class="va">start</span>, <span class="st">"-"</span>, <span class="va">heatmap_data</span><span class="op">$</span><span class="va">end</span><span class="op">)</span></pre></div>
<p>Top 50 DMRs</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span>mat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">heatmap_data</span><span class="op">[</span>,<span class="op">-</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span><span class="op">]</span>, <span class="fl">50</span><span class="op">)</span>, show_rownames <span class="op">=</span> <span class="cn">FALSE</span>, annotation_col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">meth</span><span class="op">@</span><span class="va">colData</span><span class="op">)</span>, fontsize <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></pre></div>
<p><img src="04-DMRs_files/figure-html/unnamed-chunk-20-1.png" width="336"></p>
</div>
<div id="number-and-size-of-dmrs" class="section level2">
<h2 class="hasAnchor">
<a href="#number-and-size-of-dmrs" class="anchor"></a>Number and size of DMRs</h2>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"plotly"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"plotly"</span><span class="op">)</span>
<span class="op">}</span>
<span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"ggplot2"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"ggplot2"</span><span class="op">)</span>
<span class="op">}</span>
<span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"scales"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"scales"</span><span class="op">)</span>
<span class="op">}</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://plotly-r.com">plotly</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://scales.r-lib.org">scales</a></span><span class="op">)</span></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit">

<span class="va">count</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Higher methylation in day 14"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">dmrs</span><span class="op">[</span><span class="va">dmrs</span><span class="op">$</span><span class="va">diff.Methy</span><span class="op">&lt;</span><span class="fl">0</span>,<span class="op">]</span><span class="op">)</span>,
           <span class="st">"Higher methylation in day 6"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">dmrs</span><span class="op">[</span><span class="va">dmrs</span><span class="op">$</span><span class="va">diff.Methy</span><span class="op">&gt;</span><span class="fl">0</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="va">length</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Higher methylation in day 14"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu">width</span><span class="op">(</span><span class="va">dmrs</span><span class="op">[</span><span class="va">dmrs</span><span class="op">$</span><span class="va">diff.Methy</span><span class="op">&lt;</span><span class="fl">0</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span>, 
            <span class="st">"Higher methylation in day 6"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu">width</span><span class="op">(</span><span class="va">dmrs</span><span class="op">[</span><span class="va">dmrs</span><span class="op">$</span><span class="va">diff.Methy</span><span class="op">&lt;</span><span class="fl">0</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> 


<span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>Direction<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Gain"</span>, <span class="st">"Loss"</span><span class="op">)</span>, Count<span class="op">=</span><span class="va">count</span>, Length<span class="op">=</span><span class="va">length</span><span class="op">)</span>

   <span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">data</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_col</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>, y<span class="op">=</span><span class="va">Count</span>, fill<span class="op">=</span><span class="va">Direction</span><span class="op">)</span>, position <span class="op">=</span> <span class="st">"dodge"</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>, axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_fill_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Number of differentially methylated regions"</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="va">comma</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/plotly/man/ggplotly.html">ggplotly</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></pre></div>
<div id="htmlwidget-aedd1af4c466819b1996" style="width:288px;height:288px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-aedd1af4c466819b1996">{"x":{"data":[{"orientation":"v","width":0.45,"base":0,"x":[0.775],"y":[23490],"text":"factor(1): 1<br />Count: 23490<br />Direction: Gain","type":"bar","marker":{"autocolorscale":false,"color":"rgba(27,158,119,1)","line":{"width":1.88976377952756,"color":"transparent"}},"name":"Gain","legendgroup":"Gain","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"orientation":"v","width":0.45,"base":0,"x":[1.225],"y":[13934],"text":"factor(1): 1<br />Count: 13934<br />Direction: Loss","type":"bar","marker":{"autocolorscale":false,"color":"rgba(217,95,2,1)","line":{"width":1.88976377952756,"color":"transparent"}},"name":"Loss","legendgroup":"Loss","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":40.8401826484018,"r":7.30593607305936,"b":10.958904109589,"l":60.6392694063927},"plot_bgcolor":"rgba(255,255,255,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"title":{"text":"Number of differentially methylated regions","font":{"color":"rgba(0,0,0,1)","family":"","size":17.5342465753425},"x":0,"xref":"paper"},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[0.4,1.6],"tickmode":"array","ticktext":["1"],"tickvals":[1],"categoryorder":"array","categoryarray":["1"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":false,"tickfont":{"color":null,"family":null,"size":0},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":"rgba(235,235,235,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"y","title":{"text":"","font":{"color":null,"family":null,"size":0}},"hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[-1174.5,24664.5],"tickmode":"array","ticktext":["0","5,000","10,000","15,000","20,000"],"tickvals":[0,5000,10000,15000,20000],"categoryorder":"array","categoryarray":["0","5,000","10,000","15,000","20,000"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":"rgba(235,235,235,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"x","title":{"text":"Count","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":"transparent","line":{"color":"rgba(51,51,51,1)","width":0.66417600664176,"linetype":"solid"},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":true,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895},"y":0.945866141732283},"annotations":[{"text":"Direction","x":1.02,"y":1,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"left","yanchor":"bottom","legendTitle":true}],"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","showSendToCloud":false},"source":"A","attrs":{"60a05b499e6":{"x":{},"y":{},"fill":{},"type":"bar"}},"cur_data":"60a05b499e6","visdat":{"60a05b499e6":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script><div class="sourceCode" id="cb18"><pre class="downlit">
   <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">data</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_col</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>, y<span class="op">=</span><span class="va">Length</span>, fill<span class="op">=</span><span class="va">Direction</span><span class="op">)</span>, position <span class="op">=</span> <span class="st">"dodge"</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>, axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_fill_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Length of differentially methylated regions"</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="va">comma</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/plotly/man/ggplotly.html">ggplotly</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></pre></div>
<div id="htmlwidget-24cd2e241a8b640a1e01" style="width:288px;height:288px;" class="plotly html-widget"></div>
<script type="application/json" data-for="htmlwidget-24cd2e241a8b640a1e01">{"x":{"data":[{"orientation":"v","width":0.45,"base":0,"x":[0.775],"y":[10705228],"text":"factor(1): 1<br />Length: 10705228<br />Direction: Gain","type":"bar","marker":{"autocolorscale":false,"color":"rgba(27,158,119,1)","line":{"width":1.88976377952756,"color":"transparent"}},"name":"Gain","legendgroup":"Gain","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"orientation":"v","width":0.45,"base":0,"x":[1.225],"y":[10705228],"text":"factor(1): 1<br />Length: 10705228<br />Direction: Loss","type":"bar","marker":{"autocolorscale":false,"color":"rgba(217,95,2,1)","line":{"width":1.88976377952756,"color":"transparent"}},"name":"Loss","legendgroup":"Loss","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":40.8401826484018,"r":7.30593607305936,"b":10.958904109589,"l":78.1735159817352},"plot_bgcolor":"rgba(255,255,255,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"title":{"text":"Length of differentially methylated regions","font":{"color":"rgba(0,0,0,1)","family":"","size":17.5342465753425},"x":0,"xref":"paper"},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[0.4,1.6],"tickmode":"array","ticktext":["1"],"tickvals":[1],"categoryorder":"array","categoryarray":["1"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":false,"tickfont":{"color":null,"family":null,"size":0},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":"rgba(235,235,235,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"y","title":{"text":"","font":{"color":null,"family":null,"size":0}},"hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[-535261.4,11240489.4],"tickmode":"array","ticktext":["0","3,000,000","6,000,000","9,000,000"],"tickvals":[0,3000000,6000000,9000000],"categoryorder":"array","categoryarray":["0","3,000,000","6,000,000","9,000,000"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.65296803652968,"tickwidth":0.66417600664176,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.689497716895},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":true,"gridcolor":"rgba(235,235,235,1)","gridwidth":0.66417600664176,"zeroline":false,"anchor":"x","title":{"text":"Length","font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187}},"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":"transparent","line":{"color":"rgba(51,51,51,1)","width":0.66417600664176,"linetype":"solid"},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":true,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.88976377952756,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.689497716895},"y":0.945866141732283},"annotations":[{"text":"Direction","x":1.02,"y":1,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(0,0,0,1)","family":"","size":14.6118721461187},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"left","yanchor":"bottom","legendTitle":true}],"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","showSendToCloud":false},"source":"A","attrs":{"60a0757cfdc9":{"x":{},"y":{},"fill":{},"type":"bar"}},"cur_data":"60a0757cfdc9","visdat":{"60a0757cfdc9":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.2,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<div id="region-annotation" class="section level1">
<h1 class="hasAnchor">
<a href="#region-annotation" class="anchor"></a>Region annotation</h1>
<p>To describe and interpret the differentially methylated regions, we can use the <code>ChIPseeker</code> package. The location of the DMRs can already give us a hint the involved processes, but once we assign the regions to genes, a pathway enrichment analysis is also possible. For additional details, see the <a href="https://bioconductor.org/packages/release/bioc/vignettes/ChIPseeker/inst/doc/ChIPseeker.html">vignette</a> of <code>ChIPseeker</code>:</p>
<div class="sourceCode" id="cb19"><pre class="downlit">
<span class="co">## loading packages</span>
<span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"ChIPseeker"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"ChIPseeker"</span><span class="op">)</span>
<span class="op">}</span>
<span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"TxDb.Hsapiens.UCSC.hg19.knownGene"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"TxDb.Hsapiens.UCSC.hg19.knownGene"</span><span class="op">)</span>
<span class="op">}</span>
<span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"ReactomePA"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"ReactomePA"</span><span class="op">)</span>
<span class="op">}</span>
<span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"clusterProfiler"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"clusterProfiler"</span><span class="op">)</span>
<span class="op">}</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://guangchuangyu.github.io/software/ChIPseeker">ChIPseeker</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span>
<span class="va">txdb</span> <span class="op">&lt;-</span> <span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://guangchuangyu.github.io/software/ReactomePA">ReactomePA</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://guangchuangyu.github.io/software/clusterProfiler">clusterProfiler</a></span><span class="op">)</span></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit">
<span class="va">peakAnnoList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="st">"Higher in day 14"</span><span class="op">=</span><span class="va">dmrs</span><span class="op">[</span><span class="va">dmrs</span><span class="op">$</span><span class="va">diff.Methy</span><span class="op">&lt;</span><span class="fl">0</span>,<span class="op">]</span>, <span class="st">"Higher in day 6"</span> <span class="op">=</span> <span class="va">dmrs</span><span class="op">[</span><span class="va">dmrs</span><span class="op">$</span><span class="va">diff.Methy</span><span class="op">&gt;</span><span class="fl">0</span>,<span class="op">]</span><span class="op">)</span>, <span class="va">annotatePeak</span>, TxDb<span class="op">=</span><span class="va">txdb</span>, tssRegion<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">3000</span>, <span class="fl">3000</span><span class="op">)</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>

<span class="fu">ChIPseeker</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ChIPseeker/man/plotAnnoBar-methods.html">plotAnnoBar</a></span><span class="op">(</span><span class="va">peakAnnoList</span><span class="op">)</span></pre></div>
<p><img src="04-DMRs_files/figure-html/unnamed-chunk-24-1.png" width="480"></p>
<div class="sourceCode" id="cb21"><pre class="downlit">
<span class="fu">ChIPseeker</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ChIPseeker/man/plotDistToTSS-methods.html">plotDistToTSS</a></span><span class="op">(</span><span class="va">peakAnnoList</span><span class="op">)</span></pre></div>
<p><img src="04-DMRs_files/figure-html/unnamed-chunk-24-2.png" width="480"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Anand Mayakonda, Reka Toth.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
